spring-xd-dirt/src/main/java/org/springframework/xd/dirt/module/ResourceModuleRegistry.java

```diff
@@ -16,25 +16,319 @@
+		}
+		return result;
+	}
+
+	@Override
+	public List<ModuleDefinition> findDefinitions(ModuleType type) {
+		List<ModuleDefinition> result = new ArrayList<>();
+		try {
+			for (Resource resource : getResources(type.name(), "*", "")) {
+				collect(resource, result);
+			}
+		}
+		catch (IOException e) {
+			return Collections.emptyList();
+		}
+		return result;
+	}
+
+	@Override
+	public List<ModuleDefinition> findDefinitions() {
+		List<ModuleDefinition> result = new ArrayList<>();
+		try {
+			for (Resource resource : getResources("*", "*", "")) {
+				collect(resource, result);
+			}
+		}
+		catch (IOException e) {
+			return Collections.emptyList();
+		}
+		return result;
+	}
+
+	protected Iterable<Resource> getResources(String moduleType,
+			String moduleName, String suffix) throws IOException {
+		String path = String.format("%s/%s/%s%s", this.root, moduleType, moduleName, suffix);
+		Resource[] resources = this.resolver.getResources(path);
+		List<Resource> filtered = sanitize(resources);
+
+		return filtered;
+	}
+
+	private List<Resource> sanitize(Resource[] resources) throws IOException {
+		List<Resource> filtered = new ArrayList<>();
+		for (Resource resource : resources) {
+			// Sanitize file paths (and force use of FSR, which is WritableResource)
+			if (resource instanceof UrlResource && resource.getURL().getProtocol().equals("file")
+					|| resource instanceof FileSystemResource) {
+				resource = new FileSystemResource(resource.getFile().getCanonicalFile());
+			}
+			filtered.add(resource);
+		}
+		return filtered;
+	}
+
+	protected void collect(Resource resource, List<ModuleDefinition> holder) throws IOException {
+		// ResourcePatternResolver.getResources() may return resources that don't exist when not using wildcards
+		if (!resource.exists()) {
+			return;
+		}
+
+		String path = resource.getURI().toString();
+		if (path.endsWith(HASH_EXTENSION)) {
+			return;
+		}
+		else if (path.endsWith(ARCHIVE_AS_FILE_EXTENSION) && requireHashFiles && !hashResource(resource).exists()) {
+			return;
+		}
+		// URI paths for directories include an extra slash, strip it for now
+		if (path.endsWith("/")) {
+			path = path.substring(0, path.length() - 1);
+		}
+
+		int lastSlash = path.lastIndexOf('/');
+		int nextToLastSlash = path.lastIndexOf('/', lastSlash - 1);
+
+		String name = path.substring(lastSlash + 1);
+		if (name.endsWith(ARCHIVE_AS_FILE_EXTENSION)) {
+			name = name.substring(0, name.length() - ARCHIVE_AS_FILE_EXTENSION.length());
+		}
+		String typeAsString = path.substring(nextToLastSlash + 1, lastSlash);
+		ModuleType type = null;
+		try {
+			type = ModuleType.valueOf(typeAsString);
+		}
+		catch (IllegalArgumentException e) {
+			// Not an actual type name, skip
+			return;
+		}
+		String locationToUse = resource.getURL().toString();
+
+		ModuleDefinition found = ModuleDefinitions.simple(name, type, locationToUse);
+
+		if (holder.contains(found)) {
+			SimpleModuleDefinition one = (SimpleModuleDefinition) found;
```

Comment:
The list will only contain SimpleModuleDefinitions (and we're responsible for its creation). But in java, a `List<Goose>` is not a `List<Animal>`.